### PR TITLE
Fix grouping over nonexisting keys

### PR DESCRIPTION
**Pull request by Ponewor**
There were two issues with $group stage.
1. Whenever the _id field was a dictionary itself and one of its subfields was absent in a queried document, mongomock was instantly assuming the whole _id field value is None.
2. The accumulations inside $group stage should take place even if accumulated expressions consist of non-existent fields. Instead, mongomock has been throwing an error.

It's my first time ever contributing and I'm not sure if I should have added corresponding issues using Issues panel in the first place. I hope everything is alright.

EDIT
I have added the $addFields aggregation stage and completely rewritten arithmetic operators - there were many problems with them, the most important changes are
-the unary operators now also accept single arguments given as [argument]
-$add operator supports date addition
-$subtract operator supports date and timedelta subtraction
-better handling of empty arrays and null values

### CODE REVIEW

File: tests/test__mongomock.py [hunk 1]
```diff
@@ -2489,6 +2489,68 @@ def test__aggregate_filter(self):
             'cond': {'$lt': ['$$this.price', 100]},
         }}}}])
 
+    def test__group_with_missing_fields1(self):
+        self.cmp.do.remove()
+
+        data = [
+            {'_id': ObjectId(), 'a': 0, 'b': 0},
```
Comment (pcorpet) [inline tests/test__mongomock.py]: No need to add the `_id` field it's added automatically.

File: tests/test__mongomock.py [hunk 2]
```diff
@@ -2489,6 +2489,68 @@ def test__aggregate_filter(self):
...
+            {'_id': ObjectId(), 'a': 0},
+            {'_id': ObjectId(), 'b': 0},
+            {'_id': ObjectId()}
+        ]
+        for item in data:
```
Comment (pcorpet) [inline tests/test__mongomock.py]: replace by a `self.cmp.do.insert_many(data)`

File: mongomock/aggregate.py [hunk 3]
```diff
@@ -369,7 +372,12 @@ def _accumulate_group(output_fields, group_list):
             continue
         for operator, key in six.iteritems(value):
             key_getter = functools.partial(_parse_expression, key)
```
Comment (pcorpet) [inline mongomock/aggregate.py]: you don't need the getter anymore: please inline it `values.append(_parse_expression(key, doc)`

**Review by pcorpet — CHANGES_REQUESTED** [review]
Hi @Marcin, thanks for reporting and fixing that bug.

Please rebase on top of the `develop` branch instead of merging. There's something weird happening with the CI and I hope it will solve it.

**Comment by Ponewor** [comment]
I have no idea what is wrong with Travis, I'm afraid I won't be able to fix it.

**Comment by pcorpet** [comment]
> I have no idea what is wrong with Travis, I'm afraid I won't be able to fix it.

For pylint, there were some issues on develop that I've just fixed. Fort he rest, check the logs for various runs, eg. https://travis-ci.org/mongomock/mongomock/jobs/501345726

Wow, that's a lot of changes, I'll have another look.

File: mongomock/aggregate.py [hunk 4]
```diff
@@ -178,46 +181,84 @@ def _parse_basic_expression(self, expression):
         return expression
 
     def _handle_arithmetic_operator(self, operator, values):
-        if operator == '$abs':
-            return abs(self.parse(values))
+        unary_operators = {
```
Comment (pcorpet) [inline mongomock/aggregate.py]: Please move those towards the top of the file as module constants (see `_GROUPING_OPERATOR_MAP`).

File: mongomock/aggregate.py [hunk 5]
```diff
@@ -178,46 +181,84 @@ def _parse_basic_expression(self, expression):
...
+            if isinstance(values, list):
+                if len(values) != 1:
+                    raise OperationFailure('Expression %s takes exactly 1 arguments. %d were passed in.'
+                                           % (operator, len(values)))
+                values = values[0]
+            try:
+                return None if self.parse(values) is None else unary_operators[operator](self.parse(values))
+            except KeyError:
+                return None
+        if operator in binary_operators:
```
Comment (pcorpet) [inline mongomock/aggregate.py]: Please add a blank line above to separate visually the handling of unary from the rest (same after each case please).

File: mongomock/aggregate.py [hunk 6]
```diff
@@ -178,46 +181,84 @@ def _parse_basic_expression(self, expression):
...
+            '$divide': lambda x, y: x / y,
+            '$log': math.log,
+            '$mod': math.fmod,
+            '$pow': math.pow,
+            '$subtract': lambda x, y: x - y,
+        }
+        if operator in unary_operators:
+            if isinstance(values, list):
+                if len(values) != 1:
+                    raise OperationFailure('Expression %s takes exactly 1 arguments. %d were passed in.'
```
Comment (pcorpet) [inline mongomock/aggregate.py]: Please break after `OperationFailure(` so that when you need several lines you have less indentation on your left.

File: mongomock/aggregate.py [hunk 7]
```diff
@@ -178,46 +181,84 @@ def _parse_basic_expression(self, expression):
...
+        }
+        binary_operators = {
+            '$divide': lambda x, y: x / y,
+            '$log': math.log,
+            '$mod': math.fmod,
+            '$pow': math.pow,
+            '$subtract': lambda x, y: x - y,
+        }
+        if operator in unary_operators:
+            if isinstance(values, list):
```
Comment (pcorpet) [inline mongomock/aggregate.py]: Please check for tuples as well: 
```py
if isinstance(values, (list, tuple)):
```

File: mongomock/aggregate.py [hunk 8]
```diff
@@ -178,46 +181,84 @@ def _parse_basic_expression(self, expression):
...
+                if len(values) != 1:
+                    raise OperationFailure('Expression %s takes exactly 1 arguments. %d were passed in.'
+                                           % (operator, len(values)))
+                values = values[0]
+            try:
+                return None if self.parse(values) is None else unary_operators[operator](self.parse(values))
+            except KeyError:
+                return None
+        if operator in binary_operators:
+            if not isinstance(values, list):
```
Comment (pcorpet) [inline mongomock/aggregate.py]: Same here, please accept tuples.

File: mongomock/aggregate.py [hunk 9]
```diff
@@ -178,46 +181,84 @@ def _parse_basic_expression(self, expression):
...
-            return math.log(self.parse(values))
-        if operator == '$log':
-            assert len(values) == 2, 'log must have only 2 items'
-            return math.log(self.parse(values[0]), self.parse(values[1]))
-        if operator == '$log10':
-            return math.log10(self.parse(values))
-        if operator == '$mod':
-            assert len(values) == 2, 'mod must have only 2 items'
-            return math.fmod(self.parse(values[0]), self.parse(values[1]))
+            if not isinstance(values, list):
```
Comment (pcorpet) [inline mongomock/aggregate.py]: again, check for tuples

File: mongomock/aggregate.py [hunk 10]
```diff
@@ -178,46 +181,84 @@ def _parse_basic_expression(self, expression):
...
+            for value in values:
+                try:
+                    value = self.parse(value)
+                    if value is None:
+                        return None
+                    elif isinstance(value, datetime.datetime):
+                        parsed_dates.append(value)
+                    else:
+                        parsed_numbers.append(value)
+                except KeyError:
```
Comment (pcorpet) [inline mongomock/aggregate.py]: Please reduce the try/except to the scope only where it matters (around the `self.parse(value)`).

File: mongomock/aggregate.py [hunk 11]
```diff
@@ -178,46 +181,84 @@ def _parse_basic_expression(self, expression):
...
+                    return None
+            if len(parsed_dates) > 1:
+                raise OperationFailure('only one date allowed in an $add expression')
+            result = sum(parsed_numbers)
+            if len(parsed_dates) == 1:
```
Comment (pcorpet) [inline mongomock/aggregate.py]: `if parsed_dates:` should be enough of a check as you already handled the case where there were more than 1.

File: mongomock/aggregate.py [hunk 12]
```diff
@@ -745,8 +790,18 @@ def _handle_count_stage(in_collection, database, options):
...
+            try:
+                doc[key] = _parse_expression(value, doc)
+            except KeyError:
+                pass
+    return in_collection
+
+
 _PIPELINE_HANDLERS = {
-    '$addFields': None,
+    '$addFields': _handle_add_fields_stage,
```
Comment (pcorpet) [inline mongomock/aggregate.py]: Please move that to a different MR: you already have a lot in here.

File: tests/test__mongomock.py [hunk 13]
```diff
@@ -2489,6 +2507,141 @@ def test__aggregate_filter(self):
...
+                'r': {'$add': '$non_existent_key'},
+                's': {'$multiply': [1, None]},
+                't': {'$add': [None, 1]},
+                'u': {'$multiply': ['$a', '$b', 4]},
+                'v': {'$multiply': '$b'}
+            }}
+        ]
+        self.cmp.compare.aggregate(pipeline)
+
+    def test__aggregate_exception(self):
```
Comment (pcorpet) [inline tests/test__mongomock.py]: Please add a TODO (for me) to do those comparisons more easily with a single command.

**Review by pcorpet — CHANGES_REQUESTED** [review]
I'd rather not add more and more code in a single MR. I suggest we validate some of the changes you did, and then open another MR on top of it.

File: mongomock/aggregate.py (hunk 12 above)
Comment (ronyf) [reply mongomock/aggregate.py]: We really need the `$addFields` operator... can you please push it separately?

Comment (ronyf) [reply mongomock/aggregate.py]: also, it seems like the $addFields solution doesn't support nested params, such as:
```
...
{
	"$addFields": {
		"field_a": {
			"$arrayElemAt": [
				"$field_b.field_a",
				{"$indexOfArray": ["$field_b.field_c", some_value]}
			]
		}
	}
},
...
```

File: mongomock/aggregate.py (hunk 4 above)
Comment (pcorpet) [reply mongomock/aggregate.py]: Done

File: mongomock/aggregate.py (hunk 5 above)
Comment (pcorpet) [reply mongomock/aggregate.py]: Done.

File: mongomock/aggregate.py (hunk 7 above)
Comment (pcorpet) [reply mongomock/aggregate.py]: Done.

File: mongomock/aggregate.py (hunk 8 above)
Comment (pcorpet) [reply mongomock/aggregate.py]: Done.

File: mongomock/aggregate.py (hunk 9 above)
Comment (pcorpet) [reply mongomock/aggregate.py]: Done.

File: mongomock/aggregate.py (hunk 10 above)
Comment (pcorpet) [reply mongomock/aggregate.py]: Done.

File: mongomock/aggregate.py (hunk 11 above)
Comment (pcorpet) [reply mongomock/aggregate.py]: Dropped.

File: mongomock/aggregate.py (hunk 12 above)
Comment (pcorpet) [reply mongomock/aggregate.py]: The $addFields has been added in 9b92a7e018

File: tests/test__mongomock.py (hunk 13 above)
Comment (pcorpet) [reply tests/test__mongomock.py]: Fixed.

File: mongomock/aggregate.py (hunk 6 above)
Comment (pcorpet) [reply mongomock/aggregate.py]: Done.
